pyqmc/jastrowspin.py

```diff
@@ -305,3 +304,31 @@ def pgradient(self):
         For the derivatives of basis functions, we will have to compute the derivative
         of all the b's and redo the sums, similar to recompute() """
         return {"bcoeff": self._bvalues, "acoeff": self._avalues}
+
+    def u_components(self, natom, r):
+        a_up_atm = []
+        a_dn_atm = []
+        u_onebody = {}
+        for atm in range(natom):
+            a_up = 0
+            a_dn = 0
+            for i, a in enumerate(self.a_basis):

```

Comment:
this is a bit of an anti-pattern in Python. You don't want explicit loops with lots of stuff in them. This whole thing could be two lines: evaluate the basis and then evaluate the sum.